src/Swarm/Game/Scenario/Objective/WinCheck.hs

```diff
@@ -0,0 +1,83 @@
+-- | Utilities to check whether conditions are met for a game win/loss.
+module Swarm.Game.Scenario.Objective.WinCheck where
+
+import Data.Aeson
+import Data.BoolExpr qualified as BE
+import Data.List (partition)
+import Data.Map qualified as M
+import Data.Set (Set)
+import Data.Set qualified as Set
+import GHC.Generics (Generic)
+import Swarm.Game.Scenario.Objective
+import Swarm.Game.Scenario.Objective.Logic as L
+import Swarm.Game.Scenario.Objective.Simplify qualified as Simplify
+
+-- | We have "won" if all of the "unwinnable" or remaining "incomplete" objectives are "optional".
+didWin :: ObjectiveCompletion -> Bool
+didWin oc = all _objectiveOptional $ incomplete buckets <> unwinnable buckets
+ where
+  buckets = completionBuckets oc
```

Comment:
```suggestion
didWin (completionBuckets -> buckets) = all _objectiveOptional $ incomplete buckets <> unwinnable buckets
```
(with `PatternViews` enabled).  Feel free to ignore this suggestion if you don't prefer this style.

Reply:
Hmm, I'll pass :smile: